### ISSUE TITLE
Can the adapter work with EKS Fargate type ?

### ISSUE THREAD

**Issue by rubashanti** [awslabs/k8s-cloudwatch-adapter#21]
I can see that the adapter using metadata to retrieve the region, I tried using the adapter with EKS FARGATE type cluster and I got this error :

 get hpa sqs-consumer-scaler -o yaml
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    autoscaling.alpha.kubernetes.io/conditions: '[{"type":"AbleToScale","status":"True","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"SucceededGetScale","message":"the
      HPA controller was able to get the target''s current scale"},{"type":"ScalingActive","status":"False","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"FailedGetExternalMetric","message":"the
      HPA was unable to compute the replica count: unable to get external metric default/sqs-helloworld-length/nil:
      unable to fetch metrics from external metrics API: MissingRegion: could not
      find region configuration"}]'

how can we add specific region ?

is there anything else I can do to make it work with fargate cluster

**Comment by chankh** [comment]
Try adding the environment variable `AWS_REGION` to the deployment

**Comment by rubashanti** [comment]
Can you help with this error :   

  autoscaling.alpha.kubernetes.io/conditions: '[{"type":"AbleToScale","status":"True","lastTransitionTime":"2020-01-28T00:06:41Z","reason":"SucceededGetScale","message":"the
      HPA controller was able to get the target''s current scale"},{"type":"ScalingActive","status":"False","lastTransitionTime":"2020-01-28T00:23:06Z","reason":"FailedGetExternalMetric","message":"the
      HPA was unable to compute the replica count: unable to get external metric default/sqs-helloworld-length/nil:
      unable to fetch metrics from external metrics API: the server is currently unable
      to handle the request (get sqs-helloworld-length.external.metrics.k8s.io)"},{"type":"ScalingLimited","status":"True","lastTransitionTime":"2020-01-28T00:11:45Z","reason":"TooFewReplicas","message":"the
      desired replica count is increasing faster than the maximum scale rate"}]'
    autoscaling.alpha.kubernetes.io/current-metrics: '[{"type":"External","external":{"metricName":"sqs-helloworld-length","currentValue":"0"}}]'
    autoscaling.alpha.kubernetes.io/metrics: '[{"type":"External","external":{"metricName":"sqs-helloworld-length","targetValue":"30"}}]'
    kubectl.kubernetes.io/last-applied-configuration: |

**Comment by rubashanti** [comment]
Any help here ?

**Comment by chankh** [comment]
@rubashanti did you configure IRSA for the adapter with cloudwatch permissions? can you also paste the logs from the `k8s-cloudwatch-adapter` pod?

**Comment by rubashanti** [comment]
[adapter (1).log](https://github.com/awslabs/k8s-cloudwatch-adapter/files/4133442/adapter.1.log)

**Comment by rubashanti** [comment]
The cloudwatch permissions are correctly configured

**Comment by rubashanti** [comment]
kubectl get --raw "/apis/metrics.aws/v1alpha1/namespaces/default/externalmetrics/sqs-helloworld-length/" | jq
{
  "apiVersion": "metrics.aws/v1alpha1",
  "kind": "ExternalMetric",
  "metadata": {
    "annotations": {
      "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"metrics.aws/v1alpha1\",\"kind\":\"ExternalMetric\",\"metadata\":{\"annotations\":{},\"name\":\"sqs-helloworld-length\",\"namespace\":\"default\",\"region\":\"eu-east-2\"},\"spec\":{\"env\":[{\"name\":\"AWS_REGION\",\"value\":\"us-east-2\"},{\"name\":\"region\",\"value\":\"us-east-2\"}],\"name\":\"sqs-helloworld-length\",\"queries\":[{\"id\":\"sqs_helloworld_length\",\"metricStat\":{\"metric\":{\"dimensions\":[{\"name\":\"QueueName\",\"value\":\"helloworld\"},{\"name\":\"region\",\"value\":\"us-east-2\"},{\"name\":\"AWS_REGION\",\"value\":\"us-east-2\"}],\"metricName\":\"ApproximateNumberOfMessagesVisible\",\"namespace\":\"AWS/SQS\"},\"period\":300,\"stat\":\"Average\",\"unit\":\"Count\"},\"returnData\":true}],\"resource\":{\"resource\":\"deployment\"}}}\n"
    },
    "creationTimestamp": "2020-01-28T01:03:30Z",
    "generation": 1,
    "name": "sqs-helloworld-length",
    "namespace": "default",
    "resourceVersion": "29386",
    "selfLink": "/apis/metrics.aws/v1alpha1/namespaces/default/externalmetrics/sqs-helloworld-length",
    "uid": "ffd74dd7-4169-11ea-ac40-0a4f44cf89e4"
  },
  "spec": {
    "env": [
      {
        "name": "AWS_REGION",
        "value": "us-east-2"
      },
      {
        "name": "region",
        "value": "us-east-2"
      }
    ],
    "name": "sqs-helloworld-length",
    "queries": [
      {
        "id": "sqs_helloworld_length",
        "metricStat": {
          "metric": {
            "dimensions": [
              {
                "name": "QueueName",
                "value": "helloworld"
              },
              {
                "name": "region",
                "value": "us-east-2"
              },
              {
                "name": "AWS_REGION",
                "value": "us-east-2"
              }
            ],
            "metricName": "ApproximateNumberOfMessagesVisible",
            "namespace": "AWS/SQS"
          },
          "period": 300,
          "stat": "Average",
          "unit": "Count"
        },
        "returnData": true
      }
    ],
    "resource": {
      "resource": "deployment"
    }
  }
}

**Comment by chankh** [comment]
you have to add the `AWS_REGION` as an environment variable to the pod deployment, not under the `ExternalMetric`, so it should look something like this

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: k8s-cloudwatch-adapter
  name: k8s-cloudwatch-adapter
  namespace: custom-metrics
spec:
  replicas: 1
  selector:
    matchLabels:
      app: k8s-cloudwatch-adapter
  template:
    metadata:
      labels:
        app: k8s-cloudwatch-adapter
      name: k8s-cloudwatch-adapter
    spec:
      serviceAccountName: k8s-cloudwatch-adapter
      containers:
      - name: k8s-cloudwatch-adapter
        image: chankh/k8s-cloudwatch-adapter:v0.8.0
        args:
        - /adapter
        - --cert-dir=/tmp
        - --secure-port=6443
        - --logtostderr=true
        - --v=10
        env:
        - name: AWS_REGION
          value: us-west-2
        ports:
        - containerPort: 6443
          name: https
        - containerPort: 8080
          name: http
        volumeMounts:
        - mountPath: /tmp
          name: temp-vol
      volumes:
      - name: temp-vol
        emptyDir: {}
```

**Comment by rubashanti** [comment]
Yes I added the aws_region to the env variable as you mentioned above, then i got the following error :
        HPA controller was able to get the target''s current scale"},{"type":"ScalingActive","status":"False","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"FailedGetExternalMetric","message":"the
        HPA was unable to compute the replica count: unable to get external metric
        default/sqs-helloworld-length/nil: unable to fetch metrics from external metrics
        API: NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor
        verbose messaging see aws.Config.CredentialsChainVerboseErrors"}]'

=====
I fixed it by adding AWS_KEY and AWS_SECRET in the env variables as well

no the last error is  : 
annotations:
      autoscaling.alpha.kubernetes.io/conditions: '[{"type":"AbleToScale","status":"True","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"SucceededGetScale","message":"the
        HPA controller was able to get the target''s current scale"},{"type":"ScalingActive","status":"False","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"FailedGetExternalMetric","message":"the
        HPA was unable to compute the replica count: unable to get external metric
        default/sqs-helloworld-length/nil: unable to fetch metrics from external metrics
        API: the server is currently unable to handle the request (get sqs-helloworld-length.external.metrics.k8s.io)"}]'
      autoscaling.alpha.kubernetes.io/metrics: '[{"type":"External","external":{"metricName":"sqs-helloworld-length","targetValue":"30"}}]'
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"autoscaling/v2beta1","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"name":"sqs-consumer-scaler","namespace":"default"},"spec":{"maxReplicas":10,"metrics":[{"external":{"metricName":"sqs-helloworld-length","targetValue":30},"type":"External"}],"minReplicas":1,"scaleTargetRef":{"apiVersion":"apps/v1beta1","kind":"Deployment","name":"sqs-consumer"}}}
    creationTimestamp: "2020-01-27T23:43:34Z"
    name: sqs-consumer-scaler
    namespace: default

**Comment by chankh** [comment]
can you attach the log from adapter that correspond to this message at `2020-01-27T23:43:40Z`? 

```
{"type":"ScalingActive","status":"False","lastTransitionTime":"2020-01-27T23:43:40Z","reason":"FailedGetExternalMetric","message":"the
HPA was unable to compute the replica count: unable to get external metric
default/sqs-helloworld-length/nil: unable to fetch metrics from external metrics
API: the server is currently unable to handle the request (get sqs-helloworld-length.external.metrics.k8s.io)"}```

**Comment by rubashanti** [comment]
I don't have logs at the time now, as I have removed the adapter and rebuild it with the env variables

**Comment by rubashanti** [comment]
I was looking into this log  :
======
I0131 04:45:08.034102       1 wrap.go:47] GET /apis/external.metrics.k8s.io/v1beta1/namespaces/default/sqs-helloworld-length: (3.130182573s) 400 [kube-controller-manager/v1.14.9 (linux/amd64) kubernetes/c0eccca/system:serviceaccount:kube-system:horizontal-pod-autoscaler 192.168.130.155:52864]
I0131 04:45:17.125031       1 request.go:947] Request Body: {"kind":"SubjectAccessReview","apiVersion":"authorization.k8s.io/v1beta1","metadata":{"creationTimestamp":null},"spec":{"nonResourceAttributes":{"path":"/","verb":"get"},"user":"system:anonymous","group":["system:unauthenticated"]},"status":{"allowed":false}}
I0131 04:45:17.125153       1 round_trippers.go:419] curl -k -v -XPOST  -H "Authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJjdXN0b20tbWV0cmljcyIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VjcmV0Lm5hbWUiOiJrOHMtY2xvdWR3YXRjaC1hZGFwdGVyLXRva2VuLWYyN3BnIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQubmFtZSI6Ims4cy1jbG91ZHdhdGNoLWFkYXB0ZXIiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC51aWQiOiI0NTU1ZjI1MC00MzljLTExZWEtOWZkZS0wNjA5N2MzZDJhYmUiLCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6Y3VzdG9tLW1ldHJpY3M6azhzLWNsb3Vkd2F0Y2gtYWRhcHRlciJ9.4jckRQjScyv3k8bjiw_1djkakdRGzxg0j0b32lt-wQunVwms74QMrm4rE6b3bz26fO_7KJ1ZZQYhk12wx9vT3NMpm-pTaJRR_rbtMmonecJyNpi7OouB-ZbCghdI6mjWAzYhK1iB8_KMaFV_38nWXQHssWex5ijT0KAojAm2e2t07tElq29wlhKfu63oJxjBwznwPFKppxmBIRmjZECdML5EMudSQBlLeYHoCnwUj09giIOX6kwQkqf4_idPEnFrXecPMNsXnwtRwSMIl9t5luhzJtckT0DWragojJZ4UD9Gytx5XKjvtNOPVKnwGozqeqCY2yP3KVxSKdHoj3tMSQ" -H "Accept: application/json, */*" -H "Content-Type: application/json" -H "User-Agent: adapter/v0.0.0 (linux/amd64) kubernetes/$Format" 'https://10.100.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews'
=========

what does that below mean  
user":"system:anonymous","group":["system:unauthenticated"]},"status":{"allowed":false}}
I

**Comment by chankh** [comment]
> I don't have logs at the time now, as I have removed the adapter and rebuild it with the env variables

Do you still have any issue after running the adapter with env variables?

**Comment by rubashanti** [comment]
ok the issue with the error cannot fetch metrics was during the adapter was redeploying, becuase the pod was terminated,

I thought the credential issue was solved when I used AWS_KEY AND AWS_SECRET


====
I0131 10:00:59.496283       1 request.go:947] Response Body: {"kind":"SubjectAccessReview","apiVersion":"authorization.k8s.io/v1beta1","metadata":{"creationTimestamp":null},"spec":{"nonResourceAttributes":{"path":"/","verb":"get"},"user":"system:anonymous","group":["system:unauthenticated"]},"status":{"allowed":false}}
I0131 10:00:59.496346       1 authorization.go:73] Forbidden: "/", Reason: ""
I0131 10:00:59.496406       1 wrap.go:47] GET /: (1.85209ms) 403 [Go-http-client/2.0 192.168.130.155:51578]
E0131 10:01:00.475209       1 client.go:102] err: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
E0131 10:01:00.475243       1 provider_external.go:32] bad request: NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
I0131 10:01:00.475315       1 trace.go:81] Trace[1474941318]: "List /apis/external.metrics.k8s.io/v1beta1/namespaces/default/sqs-helloworld-length" (started: 2020-01-31 10:00:57.348053136 +0000 UTC m=+111.417713526) (total time: 3.127237361s):
====
        HPA was unable to compute the replica count: unable to get external metric
        default/sqs-helloworld-length/nil: unable to fetch metrics from external metrics
        API: NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor
        verbose messaging see aws.Config.CredentialsChainVerboseErrors"}]'



 how can I inject the key and scret with the adapter ? I used this 
[....]
        args:
        - /adapter
        - --cert-dir=/tmp
        - --secure-port=6443
        - --logtostderr=true
        - --v=10
        env:
        - name: AWS_REGION
          value: us-west-1
        - name: AWS_SECRET
          value: XXXXXXXX
        - name: AWS_KEY
          value: XXXXXXX
        ports:
[....]
